Sprint 09

SelickSD opened the pull request:
- При открытии изображения на весь экран пользователь видит растянутое изображение до границ экрана. 

ВОПРОС: Изображение по настройкам растянуто, но по факту нет, как я понимаю это нормально, так как изображения в моках маленькие. Или я делаю что то не так и его тоже можно растянуть?

SelickSD committed:
Сделал экран пользователя, добавил 3 лейбла, кнопку и вьюшку для фото

SelickSD committed:
Сделал UITabBarController согласно лекии

SelickSD committed:
Сделал UINavigationController

SelickSD committed:
Сделал переход из ячейки в отдельный контроллер

SelickSD committed:
Сделали переходы к правильным изображениям

SelickSD committed:
Закончил работу по созданию перехода на изображение

SelickSD committed:
Сделал скрол вью и кнопку поделиться

SelickSD committed:
Поменял структуру данных, создал папку для Профиля
Добавил расширение для  UIColor
Сделал верстку профиля через код

helecta commented on YaImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -9,6 +9,8 @@ import UIKit
 
 class ImagesListViewController: UIViewController {
 
+    private let ShowSingleImageSegueIdentifier = "ShowSingleImage"
```
Константы принято писать с маленькой буквы, рекомендую поправить. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on YaImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -33,11 +35,26 @@ class ImagesListViewController: UIViewController {
+
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+            if segue.identifier == ShowSingleImageSegueIdentifier {
+                let viewController = segue.destination as! SingleImageViewController
```
Лучше переписать код так что бы не было force unwrapping:
```
guard let viewController = segue.destination as? SingleImageViewController,
      let indexPath = sender as? IndexPath,
      let image = UIImage(named: photosName[indexPath.row]) else { return }
```

helecta replied:
И не забывай следить за отступами. Общепринятые отступы помогают читать код быстрее. В общем случае поможет выделить весь файл cmd+A и зажать control+i.

helecta commented on YaImageFeed/ImagesList/SingleImageViewController.swift:
```diff
@@ -0,0 +1,72 @@
+import UIKit
+
+class SingleImageViewController: UIViewController {
+    @IBOutlet var backwardButton: UIButton!
```
Стоит добавить `private`
Необходимо указывать private  для того, чтобы спрятать детали реализации конкретной части функциональности, когда они используются только внутри области объявления.

Здесь и далее в проектах, не забывай проставлять `private` всем свойствам и функциям, которые не планируешь использовать вне класса.

helecta commented on YaImageFeed/ImagesList/SingleImageViewController.swift:
```diff
@@ -0,0 +1,72 @@
+
+import UIKit
+
+class SingleImageViewController: UIViewController {
```
Давай проставлять ключевое слово `final` классам для ускорения работы компилятора.
Это нужно для оптимизация работы компилятора: ты явно сообщаешь о том, 
что класс не планируется расширять наследованием. 
В таком случае комплятор будет быстрее находить адрес ячейки памяти при вызове функции.

Эта тема про диспетчеризацию, если интересно, можно почитать здесь:
https://habr.com/ru/post/673400/
https://habr.com/ru/company/simbirsoft/blog/673636/

helecta commented on YaImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,98 @@
+    private lazy var statusLabel: UILabel = {
+        let label = UILabel()
+        label.translatesAutoresizingMaskIntoConstraints = false
+        label.text = "Hello, World!"
```
Если у тебя будет приходить длинный текст, то он будет обрезаться. Чтобы это исправить, необходимо добавить `label.numberOfLines = 0`. И trailingAnchor должен быть равен `-16`

helecta commented on YaImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,98 @@
+            loginLabel.topAnchor.constraint(equalTo: view.topAnchor, constant: 180),
+            loginLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
+
+            statusLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: 282),
```
Попробуй изменить тестовый текст на более длинный(многострочный) и посмотри как будет у тебя выглядеть экран в этом случае.

Аналогичную проверку сделай и для остальных лейблов. Ты увидишь, что trailing констрейнты заданы не верно, они слишком большие у тебя и заточены под конкретный текст.

`nameLabel` и `loginLabel ` должны быть однострочными, им не надо добавлять `numberOfLines = 0`

helecta commented on YaImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -9,6 +9,8 @@ import UIKit
 
 class ImagesListViewController: UIViewController {
```
Тут тоже стоит проставить `final`

helecta commented on YaImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,98 @@
+            nameLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
+
+            loginLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: 260),
+            loginLabel.topAnchor.constraint(equalTo: view.topAnchor, constant: 180),
```
Тут лучше задавать отступ относительно `nameLabel`, а не `view`, так верстка более гибкая. К примеру, если дизайнеры изменять отступ у `nameLabel`, то в твоем варианте нужно будет менять отступы для всех элементов и высчитывать каким они стали относительно `view`.
Лучше сделать так:
`loginLabel.topAnchor.constraint(equalTo: nameLabel.bottomAnchor, constant: 8),`
И обрати внимание, что задаю относительно нижней границы `nameLabel`.

Аналогично, лучше переделать и для `nameLabel`. Его положение лучше задавать относительно `profileView`.
Положение `statusLabel` лучше задавать относительно `loginLabel`

helecta commented:
> * При открытии изображения на весь экран пользователь видит растянутое изображение до границ экрана.
> 
> ВОПРОС: Изображение по настройкам растянуто, но по факту нет, как я понимаю это нормально, так как изображения в моках маленькие. Или я делаю что то не так и его тоже можно растянуть?

Ты все сделал правильно. Дальше по проекту картинки будут загружаться из сети, они будут нормальных размеров и все будет работать как надо. Только сделай вызов rescaleAndCenterImageInScrollView, после установки minimumZoomScale и maximumZoomScale

SelickSD committed:
1 Изменил название константы, сделал имя с маленькой буквы
2 Поправил отступы, пробелы и форматирование
3 Убрал force unwrapping для функции prepare
4 Расставил уровни доступности атрибутов private
5 Расставил уровни доступности классов final
6 Изменил настройки statusLabel, добавил многострочность и указал правый край

SelickSD committed:
1. Поправил trailingAnchor для лейблов
2. Изменил привязки констрейнтов
3. Перенес кнопку назад в соответствие с макетом для SingleImageViewController и изменил ее размер
4. Поменял местами порядок вызова контроллеров для таб бара, правда сделал это заново переопределив, как их поменять на сториборде я не догадался..

SelickSD committed:
Поменял местами rescaleAndCenterImageInScrollView(image: image) и minimumZoomScale с maximumZoomScale

SelickSD committed:
Поменял цвет статус бара с черного на белый.
